test/framework/src/test/java/org/elasticsearch/test/VersionUtilsTests.java

```diff
@@ -0,0 +1,167 @@
+        assertTrue(got.onOrAfter(VersionUtils.getPreviousVersion()));
+        assertTrue(got.onOrBefore(Version.CURRENT));
+
+        // range of one
+        got = VersionUtils.randomVersionBetween(random(), VersionUtils.getFirstVersion(), VersionUtils.getFirstVersion());
+        assertEquals(got, VersionUtils.getFirstVersion());
+        got = VersionUtils.randomVersionBetween(random(), Version.CURRENT, Version.CURRENT);
+        assertEquals(got, Version.CURRENT);
+        got = VersionUtils.randomVersionBetween(random(), Version.V_6_0_0_alpha2,
+                Version.V_6_0_0_alpha2);
+        assertEquals(got, Version.V_6_0_0_alpha2);
+
+        // implicit range of one
+        got = VersionUtils.randomVersionBetween(random(), null, VersionUtils.getFirstVersion());
+        assertEquals(got, VersionUtils.getFirstVersion());
+        got = VersionUtils.randomVersionBetween(random(), Version.CURRENT, null);
+        assertEquals(got, Version.CURRENT);
+    }
+
+    static class TestPatchBranch {
+        public static final Version V_5_3_0 = Version.fromId(Version.V_5_3_0_ID);
+        public static final Version V_5_3_1 = Version.fromId(Version.V_5_3_1_ID);
+        public static final Version V_5_3_2 = Version.fromId(Version.V_5_3_2_ID);
+        public static final Version V_5_4_0 = Version.fromId(Version.V_5_4_0_ID);
+        public static final Version V_5_4_1 = Version.fromId(Version.V_5_4_1_ID);
+        public static final Version CURRENT = V_5_4_1;
+    }
+    public void testResolveReleasedVersionsForPatchBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestPatchBranch.CURRENT, TestPatchBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestPatchBranch.V_5_3_0, TestPatchBranch.V_5_3_1, TestPatchBranch.V_5_3_2, TestPatchBranch.V_5_4_0,
+                TestPatchBranch.V_5_4_1), released);
+        assertEquals(emptyList(), unreleased);
+    }
+
+    static class TestMinorBranch {
+        public static final Version V_5_3_0 = Version.fromId(Version.V_5_3_0_ID);
+        public static final Version V_5_3_1 = Version.fromId(Version.V_5_3_1_ID);
+        public static final Version V_5_3_2 = Version.fromId(Version.V_5_3_2_ID);
+        public static final Version V_5_4_0 = Version.fromId(Version.V_5_4_0_ID);
+        public static final Version CURRENT = V_5_4_0;
+    }
+    public void testResolveReleasedVersionsForMinorBranch() {
+        Tuple<List<Version>, List<Version>> t = VersionUtils.resolveReleasedVersions(TestMinorBranch.CURRENT, TestMinorBranch.class);
+        List<Version> released = t.v1();
+        List<Version> unreleased = t.v2();
+        assertEquals(Arrays.asList(TestMinorBranch.V_5_3_0, TestMinorBranch.V_5_3_1, TestMinorBranch.V_5_4_0), released);
+        assertEquals(singletonList(TestMinorBranch.V_5_3_2), unreleased);
+    }
+
+    static class TestNextBranch {
```

Comment:
This one is what I imagine 5.x looked like right after we forked the 5.4 branch.